dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/DefaultIdentifiableObjectManager.java

```diff
@@ -1239,66 +1255,40 @@ public void removeUserGroupFromSharing( String userGroupUid )
     }
 
     @SuppressWarnings( "unchecked" )
-    private <T extends IdentifiableObject> IdentifiableObjectStore<T> getIdentifiableObjectStore(
-        Class<T> type )
+    private <T extends IdentifiableObject> IdentifiableObjectStore<T> getIdentifiableObjectStore( @Nonnull T object )
     {
-        initMaps();
-
-        IdentifiableObjectStore<? extends IdentifiableObject> store = identifiableObjectStoreMap.get( type );
-
-        if ( store == null )
-        {
-            store = identifiableObjectStoreMap.get( type.getSuperclass() );
-
-            if ( store == null )
-            {
-                log.debug( "No IdentifiableObjectStore found for class: '{}'", type );
-            }
-        }
+        return getIdentifiableObjectStore( (Class<T>) HibernateProxyUtils.getRealClass( object ) );
+    }
 
-        return (IdentifiableObjectStore<T>) store;
+    @SuppressWarnings( "unchecked" )
+    private <T extends IdentifiableObject> IdentifiableObjectStore<T> getIdentifiableObjectStore(
+        @Nonnull Class<T> type )
+    {
+        return (IdentifiableObjectStore<T>) getObjectStore( type, identifiableObjectStoreCache,
+            identifiableObjectStores );
     }
 
     @SuppressWarnings( "unchecked" )
     private <T extends DimensionalObject> GenericDimensionalObjectStore<T> getDimensionalObjectStore(
-        Class<T> type )
+        @Nonnull Class<T> type )
     {
-        initMaps();
-
-        GenericDimensionalObjectStore<? extends DimensionalObject> store = dimensionalObjectStoreMap.get( type );
-
-        if ( store == null )
-        {
-            store = dimensionalObjectStoreMap.get( type.getSuperclass() );
+        return (GenericDimensionalObjectStore<T>) getObjectStore( type, dimensionalObjectStoreCache,
+            dimensionalObjectStores );
+    }
 
+    private <T extends E, E extends IdentifiableObject, S extends IdentifiableObjectStore<? extends E>> S getObjectStore(
+        Class<T> type, Map<Class<? extends E>, S> cache, Set<S> stores )
+    {
+        return cache.computeIfAbsent( type, key -> {
+            S store = stores.stream().filter( s -> s.getClazz() == key ).findFirst().orElse( null );
             if ( store == null )
             {
-                log.debug( "No DimensionalObjectStore found for class: '{}'", type );
+                // as this is within the "loader" function this will only get
+                // logged once
+                log.warn( "No IdentifiableObjectStore found for class: '{}'", type );
             }
-        }
-
-        return (GenericDimensionalObjectStore<T>) store;
+            return store;
+        } );
     }
 
-    private void initMaps()
```

Comment:
This was not thread-safe and also unnecessarily complicated. The new "cache" is a concurrent map which is initializing on demand. In a real start of the server this map should quickly fill with all stores and behave the same as a map that is pre-initialized.